### PR TITLE
Improve column labels used by TableReporter

### LINKED ISSUE #1221: TableReporter doesn't respect column label string
TableReporter doesn't respect the string that is used for the column label. Using the ConsoleReporter, I can set the name of the input and it shows up in the console output:

```
>>reporter = ConsoleReporter();
>>reporter.set_report_time_interval(0.1);
>>reporter.updInput('inputs').connect(model.getCoordinateSet.get(0).getOutput('value'), 'pin1_angle' );
>>reporter.updInput('inputs').connect(model.getCoordinateSet.get(1).getOutput('value'), 'q2' );
>>model.addComponent(reporter);
...

time|   pin1_angle|          q2|
           0|  -0.78539816|  -1.0471975|
         0.1|  -0.73474627|   -1.027509|
         0.2|  -0.60741367| -0.92328623|
         0.3|  -0.47806726| -0.61092782|
         0.4|  -0.43760328| 0.037672947|
         0.5|  -0.29961138|  0.56629776|
         0.6|  0.087742457|  0.53901707|
         0.7|   0.61300325| 0.045838744|
         0.8|      1.05583| -0.42200138|
```
However, it I use a table reporter and try to get the column name, it uses the full path instead:

```
>> coordinateReporter = TableReporter()
>> coordinateReporter.set_report_time_interval(0.1)
>> coordinateReporter.updInput('inputs').connect(model.getCoordinateSet.get(0).getOutput('value'), 'pin_angle' );
>> coordinateReporter.updInput('inputs').connect(model.getCoordinateSet.get(1).getOutput('value'), 'q2' );
...
>> table = coordinateReporter.getReport()
>> table.getNumColumns
ans = 
2
>> table.getColumnLabel(0)
ans = 
/double_pendulum/pin1/q1/value
```

### PR DESCRIPTION
**Pull request by tkuchida**
### Summary
Implements design that was proposed and approved in DevProp #1226. Fixes #1221.

### Primary changes
- Renamed "annotation" to "alias".
- Added `getShortLabel()`, `getLongLabel()`, and `setAlias()` methods.
- Updated `connect()` and `getAlias()` methods and doxygen.
- Removed existing code that created default aliases if none was provided by the user.
- Updated Reporters to use short label (ConsoleReporter) or long label (TableReporter).

### Tests added
- Created testReporters.cpp to check column headings reported by ConsoleReporter and TableReporter.
- Added tests for aliases and labels to testComponentInterface.cpp.
- Updated related Python test.

### Minor changes
- Silenced compiler warning in auxiliaryTestFunctions.h.

### CODE REVIEW

File: OpenSim/Common/ComponentConnector.h [hunk 1]
```diff
@@ -446,17 +491,9 @@ class OSIMCOMMON_API AbstractInput : public AbstractConnector {
...
-        if (leftParen != std::string::npos && rightParen != std::string::npos) {
-            annotation = connecteeName.substr(leftParen + 1,
-                                              rightParen - (leftParen + 1));
-        } else {
-            if (!channelName.empty()) {
-                annotation = channelName;
-            } else {
-                annotation = outputName;
-            }
-        }
```
Comment (chrisdembia) [inline OpenSim/Common/ComponentConnector.h]: Woohoo! Thanks for removing this!

File: OpenSim/Common/Component.h [hunk 2]
```diff
@@ -2793,9 +2793,7 @@ void Input<T>::connect(const AbstractOutput& output,
             else {
                 pathName = pathName + "/" + chan.second.getName();
             }
-            if (!annotation.empty() && annotation != chan.second.getChannelName()) {
-                pathName += "(" + annotation + ")";
-            }
+            pathName += "(" + alias + ")";
```
Comment (chrisdembia) [inline OpenSim/Common/Component.h]: I think you only want to put in the parentheses if `!alias.empty()`. Is this a change in behavior from before? See my PR review comment on `AbstractInput::parseConnecteeName()`

**Comment by chrisdembia** [comment]
(There are also two smaller items to address, as explained in the comments on the code).

File: OpenSim/Common/Component.h (hunk 2 above)
Comment (tkuchida) [inline OpenSim/Common/Component.h]: I'd prefer to have `()` if no alias has been provided by the user so there is a consistent format between aliased and non-aliased Outputs. I can change (i.e., drop the `()` if `alias.empty()`) if you prefer to have different formats.

File: OpenSim/Common/ComponentConnector.h [hunk 3]
```diff
@@ -446,17 +491,9 @@ class OSIMCOMMON_API AbstractInput : public AbstractConnector {
...
+        // Alias.
+        alias = connecteeName.substr(leftParen + 1, rightParen - (leftParen + 1));
```
Comment (tkuchida) [inline OpenSim/Common/ComponentConnector.h]: If you want me to drop `()` if `alias.empty()`, I will include checks for `npos` here.

File: OpenSim/Common/Component.h (hunk 2 above)
Comment (chrisdembia) [inline OpenSim/Common/Component.h]: I do prefer omitting the `()` but I am fine with either.

File: OpenSim/Common/Test/CMakeLists.txt [hunk 4]
```diff
@@ -15,5 +15,5 @@ OpenSimAddTests(
     TESTPROGRAMS ${TEST_PROGS}
     DATAFILES ${TEST_FILES} ${C3D_TEST_FILES} ${TRC_TEST_FILES} 
               ${MOT_TEST_FILES}
-    LINKLIBS ${BTK_LIBRARIES} osimCommon ${SIMTK_ALL_LIBS} 
+    LINKLIBS ${BTK_LIBRARIES} osimCommon osimSimulation ${SIMTK_ALL_LIBS}
```
Comment (tkuchida) [inline OpenSim/Common/Test/CMakeLists.txt]: I don't think it is sufficient to check the column labels without simulating (e.g., `initSystem()` was reconnecting the Inputs and assigning default aliases—took me a long time to find that). I'll move the test to the Simulation Tests folder.

File: OpenSim/Common/Component.h (hunk 2 above)
Comment (chrisdembia) [inline OpenSim/Common/Component.h]: If we do switch to `()` then it seems that the documentation (for parseConnecteeeName()) and (new) tests may need to be updated to reflect this? Did I miss this in the documentation/tests?

Comment (tkuchida) [inline OpenSim/Common/Component.h]: Ok, I will drop the `()` to retain existing behavior.

File: OpenSim/Common/Test/CMakeLists.txt (hunk 4 above)
Comment (chrisdembia) [inline OpenSim/Common/Test/CMakeLists.txt]: Okay; that's a good point. +1

Comment (aseth1) [inline OpenSim/Common/Test/CMakeLists.txt]: Could also invoke `TheWorld::buildUpSystem()` to invoke connecting and adding to the System in testComponentInterface. My preference would be for Reporters to be tested in Common.

File: OpenSim/Common/ComponentConnector.h (hunk 3 above)
Comment (tkuchida) [inline OpenSim/Common/ComponentConnector.h]: The checks for `npos` are already there.

File: OpenSim/Common/Test/CMakeLists.txt (hunk 4 above)
Comment (tkuchida) [inline OpenSim/Common/Test/CMakeLists.txt]: I suggest opening a new issue to augment testComponentInterface with Reporter-related tests. There are currently no tests for ConsoleReporter or TableReporter, and this PR adds tests in practical use cases, which I think is valuable.

File: OpenSim/Common/ComponentConnector.h (hunk 3 above)
Comment (chrisdembia) [inline OpenSim/Common/ComponentConnector.h]: > The checks for npos are already there.

Previously, there were checks for `npos` for the parentheses:

```
if (leftParen != std::string::npos && rightParen != std::string::npos) {
```

These have been removed.

The only check for `npos` I see now is

```
if (colon != std::string::npos) {
```

What are you referring to?

File: OpenSim/Common/Test/CMakeLists.txt (hunk 4 above)
Comment (chrisdembia) [inline OpenSim/Common/Test/CMakeLists.txt]: There are currently tests for `ConsoleReporter` and `TableReporter`, but they are of the "does it run?" kind, not of the "does it run correctly?" kind. 

The new `testReporters` is certainly valuable. A middle ground would be to view the proposed `testReporters` as a test of the functioning of Reporters with a proper Model. So we could rename that test to `testReportersWithModel` and leave it in Simulation.

**Comment by jimmyDunne** [comment]
I have made a quick PR into this branch here  #1283. Just adding a Matlab script that showcases the great changes.

File: OpenSim/Common/ComponentConnector.h (hunk 3 above)
Comment (tkuchida) [inline OpenSim/Common/ComponentConnector.h]: The checks for `npos` have not been removed. I have attached a screencap of the diff I see when I click "Files changed" at the top of the PR. The checks are also present when I pull the branch.

diff

File: OpenSim/Common/Test/CMakeLists.txt (hunk 4 above)
Comment (tkuchida) [inline OpenSim/Common/Test/CMakeLists.txt]: I like that solution: I'll rename and move to Simulation/Tests.

File: OpenSim/Common/ComponentConnector.h (hunk 3 above)
Comment (chrisdembia) [inline OpenSim/Common/ComponentConnector.h]: I now see that....I was having difficulty navigating the PR. Thanks for clarifying

File: OpenSim/Common/ComponentConnector.h [hunk 5]
```diff
@@ -551,30 +588,104 @@ class  Input : public AbstractInput {
...
+                                InputNotConnected, getName());
+        OPENSIM_THROW_IF(index >= getNumConnectees(),
+                         IndexOutOfRange,
+                         index, 0, getNumConnectees() - 1u);
+
+        const std::string alias = getAlias(index);
+        if (!alias.empty())
+            return alias;
+
+        // Assemble full path.
```
Comment (tkuchida) [inline OpenSim/Common/ComponentConnector.h]: :+1: `getPathName()` is much cleaner.

**Comment by tkuchida** [comment]
Thanks for reviewing; comments addressed.

**Comment by tkuchida** [comment]
Merged master and resolved conflicts.

Closed and re-opened PR [because AppVeyor thought the PR could not be merged](http://help.appveyor.com/discussions/problems/4830-stuck-on-appveyor-was-unable-to-build-non-mergeable).

This PR fixes #1221 and implements devprop #1226, but does not address #1287.

**Comment by tkuchida** [comment]
Test failures appear to be unrelated malfunctions on Travis/osx.

**Comment by aseth1** [comment]
Thank you @tkuchida. This PR is a significant improvement over the existing functionality and addresses a major bug (#1221).  #1287 documents potential issues and solutions that were subsequently discussed that can be addressed later. LGTM :+1: to merge.  

@chrisdembia do you have any outstanding concerns?

**Comment by chrisdembia** [comment]
> @chrisdembia do you have any outstanding concerns?

My only concern is that @jimmyDunne had created #1283 to showcase these changes in MATLAB, and I don't want us to lose track of that script! That could follow as a separate PR.

**Comment by jimmyDunne** [comment]
I will put a PR into Master once this is merged.

**Comment by aymanhab** [comment]
@jimmyDunne Will verify issues have been resolved.

**Pull request merged**